### PR TITLE
Add the option to fail hard on dynamic requires

### PR DESCRIPTION
**Pull request by ento**
plus readme improvements.

Trying to allow [dynamic requires](https://webpack.github.io/docs/context.html) like `AssetPath (name ++ ".png")` resulted in scary warnings like

> CriticalDependenciesWarning: Critical dependencies:\n2268:71-135 the request of a dependency is an expression
> ModuleNotFoundError: Module not found: Error: a dependency to an entry point is not allowed
> ModuleNotFoundError: Module not found: Error: a dependency to an entry point is not allowed
> ModuleParseError: Module parse failed: [..]elm-assets-loader/fixtures/elm-package.json Unexpected token (2:13)
> You may need an appropriate loader to handle this file type.

so this PR just adds an option to raise an error if such a dynamic require is found.

More detailed reasons described [here](https://github.com/NoRedInk/elm-assets-loader/pull/3/files#diff-168726dbe96b3ce427e7fedce31bb0bcR157).

We could detect string concatenations (`A2(_elm_lang$core$Basics_ops[\'++\'], ..)`  calls) and do fancy things, but that feels like going overboard with messing with Elm's internals, and having the guarantee that all assets can be resolved at compiled time seems to be a good thing.

### CODE REVIEW

**Comment by ento** [comment]
@rtfeldman ping - happy to do a walkthrough if that'd be helpful

File: README.md [hunk 1]
```diff
@@ -61,24 +61,44 @@ At runtime, the value of `My.Assets.star` will be something like
...
+
+  [Dynamic requires][dynamic-requires] is *not* supported. This option simply
+  controls whether or not to raise an error or skip over expressions like:
+
+  ```elm
+  example iconName =
+      AssetPath ("icon-" ++ iconName ++ ".png")
+  ```
+
+  [dynamic-requires]: https://webpack.github.io/docs/context.html#dynamic-requires
```
Comment (rtfeldman) [inline README.md:95]: ❤️ the attention to documentation detail!

File: README.md [hunk 2]
```diff
@@ -92,16 +112,23 @@ At runtime, the value of `My.Assets.star` will be something like
   module.exports = {
     ...
     elmAssetsLoader: {
-      localPath: function(path) {
-        return path.replace(/^\/public\//, "")
+      localPath: function(url) {
+        // transform `url` to a local path that resolves to a file
+        return url.replace(/^\/public\//, "")
```
Comment (rtfeldman) [inline README.md:117]: Does this ever appear like `public/` without the leading slash?

Put another way, would this be better or worse as `/^\/?public\//` instead?

File: index.js [hunk 3]
```diff
@@ -119,61 +76,108 @@ const transform = function(source, taggerName, localPath, babelOptions) {
...
 
-    if (path.node.callee.name !== taggerName) {
+    const argument = path.node.arguments[0];
+
+    // check for non-string-literals
+    if (!(t.isLiteral(argument) && typeof argument.value === 'string')) {
+      const actualCode = generate(path.node).code;
+
+      if (options.dynamicRequires === DYNAMIC_REQUIRES_ERROR) {
+        loaderContext.emitError("Failing hard to make sure all assets are run through webpack. Dynamically constructed asset path like this is not supported:" + actualCode);
```
Comment (rtfeldman) [inline index.js:135]: 😻 these error messages!

File: index.js [hunk 4]
```diff
@@ -119,61 +76,108 @@ const transform = function(source, taggerName, localPath, babelOptions) {
...
+      } else if (options.dynamicRequires === DYNAMIC_REQUIRES_WARN) {
+        loaderContext.emitWarning("This asset path, which is dynamically constructed, will not be run through webpack: " + actualCode);
+      }
+
+      /*
+        Trying to webpackify non-string-literals is like opening a can of bad things.
```
Comment (rtfeldman) [inline index.js:141]: > like opening a can of bad things

😂

**Review by rtfeldman — APPROVED** [review]
This looks excellent!

funk-overload

File: README.md (hunk 2 above)
Comment (ento) [reply README.md:117]: When you tag a path (`AssetPath "/public/images/star.png"`), the tagged string is passed as-is to `localPath`: so whether or not it comes with a leading slash or not depends solely on what kind of paths exist in the project's code base.

I'm not sure if `/^\/?public\//` is a generally copy-pasteable regex. The purpose of using a relative path like `public/` is presumably to reference an asset relative to the current path: how to translate it to a resolvable path will depend on each project's path setup.

As a usage example, I'm going to err on the side of limiting the expected input to what's in the example.

**Comment by ento** [comment]
Thanks for reviewing! merging as-is

**Pull request merged**